File: docker/hitl-ui/src/api/types/architect.ts
/**
 * Architect Board Canvas Types (P10-F01)
 * Types for the visual architecture diagramming feature
 */

/**
 * Base Excalidraw element properties (subset of full ExcalidrawElement)
 * This provides a decoupled type definition that matches Excalidraw's structure.
 * The full ExcalidrawElement type from the library is used at runtime.
 */
export interface ArchitectElement {
  id: string;
  type: string;
  x: number;
  y: number;
  width: number;
  height: number;
  strokeColor: string;
  backgroundColor: string;
  fillStyle: string;
  strokeWidth: number;
  strokeStyle: string;
  roughness: number;
  opacity: number;
  seed: number;
  version: number;
  versionNonce: number;
  isDeleted: boolean;
  groupIds: readonly string[];
  boundElements: readonly { id: string; type: string }[] | null;
  updated: number;
  link: string | null;
  locked: boolean;
  // Allow additional properties from Excalidraw
  [key: string]: unknown;
}

/**
 * Excalidraw app state properties relevant for saving/restoring
 */
export interface ArchitectAppState {
  viewBackgroundColor?: string;
  currentItemFontFamily?: number;
  currentItemFontSize?: number;
  currentItemStrokeColor?: string;
  currentItemBackgroundColor?: string;
  currentItemFillStyle?: string;
  currentItemStrokeWidth?: number;
  currentItemStrokeStyle?: string;
  currentItemRoughness?: number;
  currentItemOpacity?: number;
  zoom?: { value: number };
  scrollX?: number;
  scrollY?: number;
  // Allow additional properties from Excalidraw
  [key: string]: unknown;
}

/**
 * Represents a saved architect draft in the system
 */
export interface ArchitectDraft {
  id: string;
  name: string;
  elements: ArchitectElement[];
  appState: ArchitectAppState;
  createdAt: string;
  updatedAt: string;
}

/**
 * Represents an exported artifact from the architect canvas
 */
export interface ArchitectExport {
  draftId: string;
  format: ExportFormat;
  content: string;
  createdAt: string;
}

/**
 * Supported export formats for architect diagrams
 */
export type ExportFormat = 'svg' | 'png' | 'mmd' | 'drawio';

/**
 * Translation formats supported by the translation API
 * P10-F02 Diagram Translation
 */
export type TranslationFormat = 'png' | 'mmd' | 'drawio';

/**
 * Options for diagram translation
 * P10-F02 Diagram Translation
 */
export interface TranslateOptions {
  /** Image size for PNG output */
  size?: '2k' | '4k';
  /** Hint for Mermaid diagram type (flowchart, sequence, class, etc.) */
  diagramType?: string;
}

/**
 * Request body for the translate API
 * P10-F02 Diagram Translation
 */
export interface TranslateRequest {
  /** Raw SVG content string from Excalidraw export */
  svgContent: string;
  /** Target translation format */
  format: TranslationFormat;
  /** Optional translation options */
  options?: TranslateOptions;
}

/**
 * Response from the translate API
 * P10-F02 Diagram Translation
 */
export interface TranslateResponse {
  /** Translated content - Base64 for PNG, raw string for Mermaid/Draw.io */
  content: string;
  /** The format of the translated content */
  format: TranslationFormat;
  /** The LLM model used for translation */
  modelUsed: string;
  /** Additional metadata about the translation */
  metadata?: {
    /** Image width for PNG */
    width?: number;
    /** Image height for PNG */
    height?: number;
    /** Detected diagram type for Mermaid */
    diagramType?: string;
  };
}

/**
 * Error codes for translation failures
 * P10-F02 Diagram Translation
 */
export type TranslationErrorCode =
  | 'MODEL_ERROR'
  | 'INVALID_FORMAT'
  | 'SVG_PARSE_ERROR'
  | 'RATE_LIMIT';

/**
 * Error response from the translate API
 * P10-F02 Diagram Translation
 */
export interface TranslationError {
  /** Human-readable error message */
  error: string;
  /** Machine-readable error code */
  code: TranslationErrorCode;
  /** Additional error details */
  details?: string;
}

File: src/orchestrator/api/models/architect.py
"""Pydantic models for Architect Board translation endpoints.

This module defines the data models for translating SVG diagrams
from the Architect Board to various export formats (PNG, Mermaid, Draw.io).
"""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class ExportFormat(str, Enum):
    """Supported export formats for diagram translation."""

    PNG = "png"
    MERMAID = "mmd"
    DRAWIO = "drawio"


class TranslateRequest(BaseModel):
    """Request model for diagram translation.

    Attributes:
        svg_content: Raw SVG string from Excalidraw export.
        format: Target export format.
        options: Optional format-specific options.
            - For PNG: {"width": int, "height": int}
            - For Mermaid: {"diagram_type": str}  # e.g., "flowchart", "sequence"
            - For Draw.io: {"style": str}
    """

    svg_content: str = Field(
        ...,
        alias="svgContent",
        description="Raw SVG string from Excalidraw export",
        min_length=1,
    )
    format: ExportFormat = Field(
        ...,
        description="Target export format",
    )
    options: dict[str, Any] | None = Field(
        default=None,
        description="Format-specific options (size for PNG, diagramType hint for Mermaid)",
    )

    model_config = {"populate_by_name": True}


class TranslateResponse(BaseModel):
    """Response model for successful diagram translation.

    Attributes:
        content: Translated content (Base64 for PNG, raw string for Mermaid/Draw.io).
        format: The format of the translated content.
        model_used: The LLM model used for translation (for transparency).
        metadata: Optional format-specific metadata.
            - For PNG: {"width": int, "height": int, "size_bytes": int}
            - For Mermaid: {"diagram_type": str}
            - For Draw.io: {"element_count": int}
    """

    content: str = Field(
        ...,
        description="Base64 for PNG, raw string for Mermaid/Draw.io",
    )
    format: ExportFormat = Field(
        ...,
        description="The format of the translated content",
    )
    model_used: str = Field(
        ...,
        serialization_alias="modelUsed",
        description="The LLM model used for translation",
    )
    metadata: dict[str, Any] | None = Field(
        default=None,
        description="Format-specific metadata (width/height for PNG, diagramType for Mermaid)",
    )

    model_config = {"populate_by_name": True}


class TranslateErrorCode(str, Enum):
    """Error codes for translation failures."""

    MODEL_ERROR = "MODEL_ERROR"
    INVALID_FORMAT = "INVALID_FORMAT"
    SVG_PARSE_ERROR = "SVG_PARSE_ERROR"
    RATE_LIMIT = "RATE_LIMIT"
    AGENT_NOT_CONFIGURED = "AGENT_NOT_CONFIGURED"
    UNSUPPORTED_FORMAT = "UNSUPPORTED_FORMAT"


class TranslateErrorResponse(BaseModel):
    """Response model for translation errors.

    Attributes:
        error: Human-readable error message.
        code: Machine-readable error code.
        details: Optional additional error details.
    """

    error: str = Field(
        ...,
        description="Human-readable error message",
    )
    code: TranslateErrorCode = Field(
        ...,
        description="Machine-readable error code",
    )
    details: str | None = Field(
        default=None,
        description="Optional additional error details",
    )

    model_config = {"populate_by_name": True}
